ipn/ipnlocal/local.go

```diff
@@ -606,6 +607,53 @@ func (b *LocalBackend) GetDNSOSConfig() (dns.OSConfig, error) {
 	return manager.GetBaseConfig()
 }
 
+// QueryDNS performs a DNS query for name and queryType using the built-in DNS resolver, and returns
+// the raw DNS response and the resolvers that are were able to handle the query (the internal forwarder
+// may race multiple resolvers).
+func (b *LocalBackend) QueryDNS(name string, queryType dnsmessage.Type) (res []byte, resolvers []*dnstype.Resolver, err error) {
+	manager, ok := b.sys.DNSManager.GetOK()
+	if !ok {
+		return nil, nil, errors.New("DNS manager not available")
+	}
+	if !strings.HasSuffix(name, ".") {
+		name += "."
+	}
+	n, err := dnsmessage.NewName(name)
+	if err != nil {
+		b.logf("DNSQuery: failed to parse name %q: %v", name, err)
+		return nil, nil, err
+	}
+	from := netip.MustParseAddrPort("127.0.0.1:0")
+	db := dnsmessage.NewBuilder(nil, dnsmessage.Header{
+		OpCode:           0,
+		RecursionDesired: true,
+		ID:               1,
+	})
+	db.StartQuestions()
+	db.Question(dnsmessage.Question{
+		Name:  n,
+		Type:  queryType,
+		Class: dnsmessage.ClassINET,
+	})
+	q, err := db.Finish()
+	if err != nil {
+		b.logf("DNSQuery: failed to build query: %v", err)
+		return nil, nil, err
+	}
+	res, err = manager.Query(b.ctx, q, "udp", from)
```

Comment:
Slightly nitpicky, but we should probably use the `"tcp"` address type here, or make it configurable, to work around truncation issues. We handle query truncation issues by falling back from UDP to TCP, but only if the input query is itself TCP (since we'd then truncate the response). So, by using `"tcp"` here, we should be able to return the full, un-truncated query in all cases. See:
https://github.com/tailscale/tailscale/blob/main/net/dns/resolver/forwarder.go#L560-L567

Reply:
Done